lakefile.lean

```diff
@@ -61,6 +61,26 @@ lean_lib Cache where
 lean_exe cache where
   root := `Cache.Main
 
+/--
+When a package depending on Mathlib updates its dependencies,
+update its toolchain to match Mathlib's and fetch the new cache.
+-/
+post_update pkg do
+  let rootPkg ← getRootPackage
+  if rootPkg.name = pkg.name then
+    return -- do not run in Mathlib itself
+  let wsToolchainFile := rootPkg.dir / "lean-toolchain"
+  let mathlibToolchain ← IO.FS.readFile <| pkg.dir / "lean-toolchain"
+  IO.FS.writeFile wsToolchainFile mathlibToolchain
```

Comment:
Isn't this going to interfere with any eventual Lake feature to do the same thing, or with the vscode plugin, or with an unrelated package that declares itself special and decides to do the same thing?

Reply:
Yeah, but currently Mathlib requires all downstream packages be on its toolchain (for cache to work) so if any disagree occurs due to any of these other cases things would break anyway. This is designed to be a short-term mathlib-specific solution, partly as test of this kind of feature.

Reply:
Can you add a TODO comment to that effect, referencing the lean4/lake issue?

Reply:
@mhuisi, can you comment on whether this is problematic for the extension?

Reply:
Having the GUI fight against the tool like that sounds pretty confusing :(

Reply:
My proposal in #2752 involved prompting the user with a choice on the command line in the face of ambiguity. If that were implemented, then the GUI should just act as a proxy for that choice (including not asking at all in the common case where there is no ambiguity).

Reply:
@eric-wieser The problem is that, from @mhuisi's explanation, the extensions chooses to update the toolchain before the `lake update` is run (to run the right version of Lake). Thus, with leanprover/lean4#2752, either the user says yes and everything works regardless (as with this PR) or the user says no and then `lake update` updates the toolchain anyway in the case of no ambiguity. This PR is essentially producing the behavior that PR would have with a `tracks_toolchain` on Mathlib. 

From the discussion, if we wish to keep the user option to forbid an automatic update in the VS Code extensions it seems like the following 3-part approach may be best:
1. Have the VS Code extension invoke a Lake command (possibly `lake update` itself) that resolve dependencies and tells the VSCode when it needs a toolchain update and what the candidate toolchains are (but does not ever update them itself).
2. VSCode presents the user with a selection of the viable toolchains or just a button confirming whether to update in the unambiguous case.
3. VSCode invokes another Lake command (possibly also just `lake update`)  to complete the update and run post-update hooks.

Reply:
If we have decided that updating the toolchain *after* running Lake is fine (which this PR seems to suggest), then I could also ask the user whether they want to update the toolchain (if it can be updated) after `lake update <dep>`. 

In this case, the user will never be asked for Mathlib as Lake already updates the toolchain for us, but will potentially be asked for projects that do not have this post-update hook.

Reply:
@eric-wieser From this, I think the central question is whether this is fine as short-term solution while a more principled long term solution is fleshed out.

Reply:
@mhuisi

> If we have decided that updating the toolchain _after_ running Lake is fine


Do you mean "it's fine to do it after *instead of* before" or "its fine to do it after *as well as* before". I'm fine with a short-to-medium term hack here as long as it doesn't result in weird vscode behavior, which it would do if we merged this without a corresponding extension change.

Reply:
@eric-wieser / @mhuisi A potentially simple (backwards-compatible) fix would be to just drop the dialogue option for now and update without asking (since that is what this PR will do), then there would not longer be a clash.

Reply:
I don't think it makes sense to wait on this PR for changes to the extension.

The current behaviour (`lake update` leaves you with a completely broken project, and no hint that you need to change your `lean-toolchain`) is clearly much worse than the situation we will get from merging now (`lake update` leaves you in a sensible state, and the extension's behaviour is annoying).

Of course we should follow up with further changes to the extension.

Reply:
Sorry, I should have said "a corresponding *planned* extension change".

Of course, `lake update` still leaves you with a completely broken project if you're a CS person who is scared of Mathlib and using only Std, so arguably we should really be focusing on fixing things upstream (or duplicate this patch in Std...).

Reply:
@eric-wieser From my understanding, Std does not wed you to its toolchain, and one of the future goals of the project is to be compatible is to be compatible with multiple toolchain versions. Thus, it is not clear if it wants to have users to closely track its toolchain.

Reply:
Regardless, @eric-wieser. if you are happy with this change and with planned ideas for the extension, is it fine to resolve this conversation?